internal/mirror/mirror.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mirror
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/google/exposure-notifications-server/internal/database"
+	"github.com/google/exposure-notifications-server/internal/mirror/model"
+	"github.com/google/exposure-notifications-server/internal/storage"
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"go.opencensus.io/trace"
+)
+
+const mirrorLockPrefix = "mirror-lock"
+
+type FileStatus struct {
+	Order        int
+	MirrorFile   *model.MirrorFile
+	DownloadPath string
+	Filename     string
+	Failed       bool
+	Saved        bool
+}
+
+func (f *FileStatus) needsDelete() bool {
+	return f.DownloadPath == ""
+}
+
+func (f *FileStatus) needsDownload() bool {
+	return f.MirrorFile == nil
+}
+
+func (s *Server) downloadIndex(mirror *model.Mirror) ([]string, error) {
+	client := &http.Client{
+		Timeout: s.config.IndexFileDownloadTimeout,
+	}
+	resp, err := client.Get(mirror.IndexFile)
+	if err != nil {
+		return nil, fmt.Errorf("error downloading index file: %w", err)
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
```

Comment:
ReadAll has no limit and will consume all available memory given. We should use a limitReader or something with a sane max here.

Reply:
added limits to both downloads